src/handlers/users.rs

```diff
@@ -0,0 +1,73 @@
+use warp::Filter;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+use warp::reply::Reply;
+use std::str::FromStr;
+use crate::users::user::{UserDB, UserRegistration, UserCredentials, SessionToken, Session};
+use crate::error::Result;
+
+type DB<T> = Arc<RwLock<T>>;
+
+pub fn authenticate<T: UserDB>(user_db: DB<T>) -> impl warp::Filter<Extract = (Session,), Error = warp::Rejection> + Clone {
+    async fn do_authenticate<T: UserDB>(user_db: DB<T>, token: String) -> Result<Session, warp::Rejection>  {
```

Comment:
in general I agree that it's nice to keep tests close to the code. But here a lot of steps build up on each other. to test the lgoout I first have to register and then login to finally log out. this would mean a lot of code duplication.

Reply:
Maybe you do not need to use all Http calls to create a logged in a scenario in which you have to do one logout call. But if this does not work out, you could at least put the tests in this file as a submodule.

Reply:
8a9b7b5 but as I said this resulted in some code duplication